kythe/cxx/indexer/cxx/IndexerASTHooks.cc

```diff
@@ -3827,856 +3986,606 @@ void IndexerASTVisitor::DumpTypeContext(unsigned Depth, unsigned Index) {
-      }
-    }; break;
-    case TypeLoc::ObjCTypeParam: {
-      const auto& OTPTL = TypeLoc.castAs<ObjCTypeParamTypeLoc>();
-      const auto* OTPT = dyn_cast<ObjCTypeParamType>(PT);
-
-      if (const auto* TPD = OTPT ? OTPT->getDecl() : OTPTL.getDecl()) {
-        ID = BuildNodeIdForDecl(TPD);
-      }
-
-    } break;
-    case TypeLoc::BlockPointer: {
-      const auto& BPTL = TypeLoc.castAs<BlockPointerTypeLoc>();
-      const auto& DT = dyn_cast<BlockPointerType>(PT);
-      InEmitRanges = IndexerASTVisitor::EmitRanges::No;
-      auto PointeeID(BuildNodeIdForType(BPTL.getPointeeLoc(),
-                                        DT ? DT->getPointeeType().getTypePtr()
-                                           : BPTL.getPointeeLoc().getTypePtr(),
-                                        EmitRanges));
-      if (!PointeeID) {
-        return PointeeID;
-      }
-      if (SR.isValid() && SR.getBegin().isFileID()) {
-        SR.setEnd(GetLocForEndOfToken(*Observer.getSourceManager(),
-                                      *Observer.getLangOptions(),
-                                      BPTL.getCaretLoc()));
-      }
-      if (TypeAlreadyBuilt) {
-        break;
-      }
-      ID = ApplyBuiltinTypeConstructor("ptr", PointeeID);
-    } break;
       UNSUPPORTED_CLANG_TYPE(Atomic);
       UNSUPPORTED_CLANG_TYPE(Pipe);
-      // Attributed is handled in a special way at the top of this function so
-      // it is impossible for the typeloc to be TypeLoc::Attributed.
-      UNSUPPORTED_CLANG_TYPE(Attributed);
-      UNSUPPORTED_CLANG_TYPE(DependentAddressSpace);
       UNSUPPORTED_CLANG_TYPE(DependentVector);
-  }
-  if (TypeAlreadyBuilt) {
-    ID = Prev->second;
-  } else {
-    TypeNodes[Key] = ID;
-  }
-  if (SR.isValid() && SR.getBegin().isFileID() &&
-      InEmitRanges == IndexerASTVisitor::EmitRanges::Yes) {
-    if (auto RCC =
-            ExplicitRangeInCurrentContext(ExpandRangeIfEmptyFileID(SR))) {
-      if (const auto& SpellID = (SpanID ? SpanID : ID)) {
-        Observer.recordTypeSpellingLocation(*RCC, *SpellID, Claimability,
-                                            IsImplicit(*RCC));
-      }
     }
-  }
-  return ID;
+#undef UNSUPPORTED_CLANG_TYPE
+#undef DELEGATE_TYPE
+  }();
+  return (TypeNodes[Key] = Nodes);
 }
 
 absl::optional<GraphObserver::NodeId>
-IndexerASTVisitor::RecordObjCInterfaceType(
-    GraphObserver::NodeId BaseType, bool BaseIsId, const ObjCObjectType* T,
-    ArrayRef<SourceLocation> ProtocolLocs) {
-  const auto Protocols = T->getProtocols();
-  if (Protocols.empty()) {
-    return BaseType;
+IndexerASTVisitor::BuildNodeIdForObjCProtocols(clang::ObjCObjectTypeLoc TL) {
+  if (TL.getTypePtr()->getInterface()) {
+    if (auto BaseId = BuildNodeIdForType(TL.getBaseLoc())) {
+      return BuildNodeIdForObjCProtocols(*BaseId, TL.getTypePtr());
+    } else {
+      return absl::nullopt;
+    }
+  } else {
+    return BuildNodeIdForObjCProtocols(TL.getTypePtr());
   }
+}
+
+// Base case.
```

Comment:
What is this the base case for?

Reply:
I found the previous merged logic harder to follow with the bool `BaseIsId`; the comment was originally just to aid my thinking. I've expanded it.